File: string-to-integer-atoi/shaun-gy-string-math.cpp
//grandyang
//shaun read and understand
//runtime 4ms, faster than 90%, mem less than 46%
class Solution {
public:
    int myAtoi(string str) {
        if (str.empty()) return 0;
        
        int sign = 1, base = 0, i = 0, n = str.size();
        //skip leading space
        while (i < n && str[i] == ' ') ++i;
        
        //deal with + - before digits
        if (i < n && (str[i] == '+' || str[i] == '-')) {
            sign = (str[i++] == '+') ? 1 : -1;
        }
        
        
        while (i < n && str[i] >= '0' && str[i] <= '9') {
            //dealing with number(base) out of range of 32bit
            //  INT_MAX/10 * 10 will be + (str[i] - '0')
            if (base > INT_MAX / 10 || (base == INT_MAX / 10 && str[i] - '0' > 7)) {
                return (sign == 1) ? INT_MAX : INT_MIN;
            }
            base = 10 * base + (str[i++] - '0');
        }
        
        return base * sign;
    }
};

File: string-to-integer-atoi/shaun-gy-string-math.java
/*
Runtime: 1 ms, faster than 100.00% of Java online submissions for String to Integer (atoi).
Memory Usage: 39.4 MB, less than 5.59% of Java online submissions for String to Integer (atoi).
Next challenges:


*/
class Solution {
    public int myAtoi(String str) {
        if (str.length() == 0) return 0;

        int sign = 1, base = 0, i = 0, n = str.length();
        //skip leading space
        while (i < n && str.charAt(i) == ' ') ++i;

        //deal with +/- before digits
        if (i < n && (str.charAt(i) == '+' || str.charAt(i) == '-')) {
            sign = (str.charAt(i) == '+') ? 1 : -1;
            i++; //move to next char
        }


        while (i < n && str.charAt(i) >= '0' && str.charAt(i) <= '9') {
            //dealing with number(base) out of range of 32bit
            //  INT_MAX/10 * 10 will be + (str[i] - '0')
            if (base > Integer.MAX_VALUE / 10 || 
                (base == Integer.MAX_VALUE / 10 && str.charAt(i) - '0' > 7)) {
                return (sign == 1) ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }
            base = 10 * base + (str.charAt(i) - '0');
            i++;
        }

        return base * sign;

    }
}
